File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        // 2 1 5 4 3 0 0 , 2 3 0 0 1 4 5
        int bp = -1;
        for(int i=nums.length-1; i>0; i--) {
            if(nums[i] > nums[i-1]) {
                bp = i-1;
                break;
            }
        }

        if(bp != -1) {
            int nextGreater = -1;
            for(int i=nums.length-1; i>bp; i--) {
                if(nums[bp] < nums[i]) {
                    nextGreater = i;
                    break;
                }
            }

            int temp = nums[bp];
            nums[bp] = nums[nextGreater];
            nums[nextGreater] = temp;
        }

        reverse(bp+1, nums.length-1, nums);
    }

    public void reverse(int l, int r, int[] nums) {
        while(l<r) {
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }
    }
}

File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        # 2 1 5 4 3 0 0 , 2 3 0 0 1 4 5

        n = len(nums)
        bp = -1
        
        for i in range(n-1,-1,-1) :
            if nums[i]>nums[i-1]:
                bp = i-1
                break

        if bp != -1:
            nextGreater = -1
            for i in range(n-1, bp, -1):
                if nums[i] > nums[bp]:
                    nextGreater = i
                    break

            nums[bp], nums[nextGreater] = nums[nextGreater], nums[bp]

        self.reverse(nums, bp+1, n-1)

    def reverse(self, nums, l, r):
        while l<r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
    
        